internal/zinc-core/src/main/scala/sbt/internal/inc/Compile.scala

```diff
@@ -176,6 +174,12 @@ private final class AnalysisCallback(
   private[this] val binaryClassName = new HashMap[File, String]
   // source files containing a macro def.
   private[this] val macroClasses = Set[String]()
+  private[this] val prevJar = {
+    JarUtils
+      .getOutputJar(output)
+      .filter(_.exists())
+      .map(_ => JarUtils.createPrevJarPath())
```

Comment:
We're assuming here that the output jar doesn't change across different compiler iterations, right? We probably need to add this invariant to the scaladoc of `Output` or output directories so that the build tool knows that it needs to guarantee that the jar will always be the same and, if it isn't, incremental compilation will not work correctly.

Reply:
If you mean between zinc runs - it can change. it can change exactly the contents of output directory. We will collect stamps at the beginning and if something was changed or deleted it will be invalidated.
The prev jar path that is generated here is done per scalac run. This happens because we create new callback for each scalac run and this is intended. So between iterations prev jar (both contents and location) and output jar (only contents) will change.
Imagine we have no output yet, we run zinc, it will just need one cycle and will produce an output jar. Then, we change some stuff and run zinc again. It will generate a random path in temp dir in analysis callback, then it will move output jar to that path and produce new output jar (that is small, contains only output from the first cycle of incremental compilation, say we recompiled 5 sources, we got jar with 5 classes). Then, those new classes are appended to previous jar, and previous jar is moved back to be the output jar. If we need another cycle, we move the modified output jar to a new prev jar path, compile some classes and merge. And so on.

I am not sure if I understood the question correctly, but I hope I answered it :)

Reply:
I was referring here to the output jar (the one specified by `output` and which needs to be a jar so that straight-to-jar compilation is enabled). I figured it would be useful to tell the build tool that the location of the output jar has to be consistently the same, but now that I think about it more this is obvious (and it's an invariant that applies to output dir being a directory instead of a jar), so forget my initial suggestion.